File: Recursion and Backtracking/Graph Coloring/GraphColoring.java
/***
 * Code Written By: Nishkarsh Makhija
 * Code Written for: Learn Algorithm
 * Insta Handle : learn_algorithm  
**/

class GraphColoring{
    
	// Number of vertices in the graph
    static final int V = 4;
    
    static boolean isSafe(int v, boolean[][] graph, int[] color, int c) {
		// Check whether adjacent vertices has same colors
        for (int i = 0; i < V; i++){
            if (graph[v][i] && c == color[i]){
                return false;
			}
		}
        return true;
    }
    
    public static boolean graphColoring(boolean[][] graph,int m,int[] color,int v){
		
		// If all vertices are assigned some color, return true
        if(v == V)
            return true;

		// For current vertex v, try different colors c
        for(int c = 1; c <= m; c++) {
			
			// Check if it is safe to assign color c to vertex v
            if (isSafe(v, graph, color, c)) {
                color[v] = c;
				
				// Go to next vertex
                if (graphColoring(graph, m, color, v + 1))
                    return true;
				
				// Backtrack
                color[v] = 0;
            }
        }

		// If no color can be assigned to vertex v, return False
        return false;
    }
    
    public static void main(String args[]){
		// Adjacent Matrix to represent a graph
        boolean[][] graph = {
            { false, true, true, true },
            { true, false, true, false },
            { true, true, false, true },
            { true, false, true, false }
        };
        
		// Color array to store color for particular vertex
        int[] color = new int[V];
        for (int i = 0; i < V; i++)
            color[i] = 0;

        // Number of colors
        int m = 3;

        if(graphColoring(graph, m, color, 0)){
            for (int i = 0; i < V; i++)
                System.out.print(" " + color[i] + " ");
            System.out.println();
        }
        else{
            System.out.println("No Solution");
        }
    }
}

File: Recursion and Backtracking/Graph Coloring/Graph Coloring.py
'''
 * Code Written By: Nishkarsh Makhija
 * Code Written for: Learn Algorithm
 * Insta Handle : learn_algorithm  
'''

# Number of vertices in the graph
V = 4

def is_safe(v,graph,color,c):
    # Check whether adjacent vertices has same colors
    for i in range(V):
        if(graph[v][i] and color[i]==c):
            return False
    
    return True

def graph_coloring(graph, m, color, v):
    
    # If all vertices are assigned some color, return true
    if (v == V):
        return True
    
    # For current vertex v, try different colors c
    for c in range (1,m+1):
     
        # Check if it is safe to assign color c to vertex v
        if is_safe(v,graph,color,c):
            color[v] = c
            
            # Go to next vertex
            if graph_coloring(graph,m,color,v+1):
                return True
                
            # Backtrack
            color[v] = 0
    
    # If no color can be assigned to vertex v, return False
    return False
    

def main():
    # Adjacent Matrix to represent a graph
    graph = [
        [0, 1, 1, 1],
        [1, 0, 1, 0],
        [1, 1, 0, 1],
        [1, 0, 1, 0],
    ]
    
    # Color array to store color for particular vertex
    color = [0] * V

    # Number of colors
    m = 3
    
    if graph_coloring(graph, m, color, 0):
        print(color)
    else:
        print("No Solution")
    
main()



